parallel/slice.go

```diff
@@ -1,98 +1,56 @@
 package parallel
 
-import "sync"
+import (
+	"runtime"
+	"sync"
+)
 
 // Map manipulates a slice and transforms it to a slice of another type.
-// `iteratee` is call in parallel. Result keep the same order.
+// `iteratee` is called in parallel. Result keep the same order.
 func Map[T any, R any](collection []T, iteratee func(T, int) R) []R {
 	result := make([]R, len(collection))
-
-	var wg sync.WaitGroup
-	wg.Add(len(collection))
-
-	for i, item := range collection {
-		go func(_item T, _i int) {
-			res := iteratee(_item, _i)
-
-			result[_i] = res
-
-			wg.Done()
-		}(item, i)
-	}
-
-	wg.Wait()
-
+	NewTaskPool(collection, runtime.NumCPU()/2, func(v T, i int) {
+		result[i] = iteratee(v, i)
+	})
 	return result
 }
 
 // ForEach iterates over elements of collection and invokes iteratee for each element.
-// `iteratee` is call in parallel.
+// `iteratee` is called in parallel.
 func ForEach[T any](collection []T, iteratee func(T, int)) {
-	var wg sync.WaitGroup
-	wg.Add(len(collection))
-
-	for i, item := range collection {
-		go func(_item T, _i int) {
-			iteratee(_item, _i)
-			wg.Done()
-		}(item, i)
-	}
-
-	wg.Wait()
+	NewTaskPool(collection, runtime.NumCPU()/2, iteratee)
 }
 
 // Times invokes the iteratee n times, returning an array of the results of each invocation.
 // The iteratee is invoked with index as argument.
-// `iteratee` is call in parallel.
+// `iteratee` is called in parallel.
 func Times[T any](count int, iteratee func(int) T) []T {
 	result := make([]T, count)
-
-	var wg sync.WaitGroup
-	wg.Add(count)
-
-	for i := 0; i < count; i++ {
-		go func(_i int) {
-			item := iteratee(_i)
-
-			result[_i] = item
-
-			wg.Done()
-		}(i)
-	}
-
-	wg.Wait()
-
+	NewTaskPool[int](make([]int, count), runtime.NumCPU()/2, func(_ int, i int) {
+		result[i] = iteratee(i)
+	})
 	return result
 }
 
 // GroupBy returns an object composed of keys generated from the results of running each element of collection through iteratee.
 // `iteratee` is call in parallel.
 func GroupBy[T any, U comparable](collection []T, iteratee func(T) U) map[U][]T {
 	result := map[U][]T{}
+	resultMu := &sync.Mutex{}
 
-	var mu sync.Mutex
-	var wg sync.WaitGroup
-	wg.Add(len(collection))
-
-	for _, item := range collection {
-		go func(_item T) {
-			key := iteratee(_item)
-
-			mu.Lock()
-
-			if _, ok := result[key]; !ok {
-				result[key] = []T{}
-			}
-
-			result[key] = append(result[key], _item)
+	NewTaskPool[T](collection, runtime.NumCPU()/2, func(v T, _ int) {
+		key := iteratee(v)
 
-			mu.Unlock()
-			wg.Done()
-		}(item)
-	}
-
-	wg.Wait()
+		resultMu.Lock()
```

Comment:
I wonder if we could remove this lock as well, pipelining results into a channel and a single consumer.

I don't know the cost of mutex versus channel. Can you execute a benchmark?

Or a simpler implementation would be to write `iteratee` result into a slice and group later?

Reply:
Tried both. This is the fastest way for now. Unless we get to know the groups upfront, this is the fastest way.